sys/suit/v4/cbor.c

```diff
@@ -0,0 +1,280 @@
+ * directory for more details.
+ */
+/**
+ * @ingroup     sys_suit
+ * @{
+ *
+ * @file
+ * @brief       SUIT manifest parser library for CBOR based manifests
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+#include <string.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#include "suit/v4/handlers.h"
+#include "suit/v4/suit.h"
+#include "suit/v4/policy.h"
+#include "cbor.h"
+#include "cose/sign.h"
+
+#include "public_key.h"
+
+#include "log.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+
+static suit_manifest_handler_t _manifest_get_auth_wrapper_handler(int key);
+typedef suit_manifest_handler_t (*suit_manifest_handler_getter_t)(int key);
+
+int suit_cbor_map_iterate_init(CborValue *map, CborValue *it)
+{
+    if (!cbor_value_is_map(map)) {
+        LOG_INFO("suit_v4_parse(): manifest not an map\n)");
+        return SUIT_ERR_INVALID_MANIFEST;
+    }
+
+    cbor_value_enter_container(map, it);
+
+    return SUIT_OK;
+}
+
+int suit_cbor_map_iterate(CborValue *it, CborValue *key, CborValue *value)
+{
+    if (cbor_value_at_end(it)) {
+        return 0;
+    }
+
+    *key = *it;
+    cbor_value_advance(it);
+
+    *value = *it;
+    cbor_value_advance(it);
+
+    return 1;
+}
+
+int suit_cbor_get_int(const CborValue *it, int *out)
+{
+    if (!cbor_value_is_integer(it)) {
+        LOG_DEBUG("expected integer type, got %u\n", cbor_value_get_type(it));
+        return SUIT_ERR_INVALID_MANIFEST;
+    }
+
+    /* This check tests whether the integer fits into "int", thus the check
+     * is platform dependent. This is for lack of specification of actually
+     * allowed values, to be made explicit at some point. */
+    if (cbor_value_get_int_checked(it, out) == CborErrorDataTooLarge) {
+        LOG_DEBUG("integer doesn't fit into int type\n");
+        return SUIT_ERR_INVALID_MANIFEST;
+    }
+
+    return SUIT_OK;
+}
+
+int suit_cbor_get_string(const CborValue *it, const uint8_t **buf, size_t *len)
+{
+    if (!(cbor_value_is_text_string(it) || cbor_value_is_byte_string(it) || cbor_value_is_length_known(it))) {
+        return SUIT_ERR_INVALID_MANIFEST;
+    }
+    CborValue next = *it;
+    cbor_value_get_string_length(it, len);
+    cbor_value_advance(&next);
+    *buf = next.ptr - *len;
+    return SUIT_OK;
+}
+
+int suit_cbor_get_uint32(const CborValue *it, uint32_t *out)
+{
+    int res;
+    int64_t val;
+    if (!cbor_value_is_unsigned_integer(it)) {
+        return CborErrorIllegalType;
+    }
+    if ((res = cbor_value_get_int64_checked(it, &val))) {
+        return res;
+    }
+    if (val > 0xFFFFFFFF) {
+        return CborErrorDataTooLarge;
+    }
+    *out = (val & 0xFFFFFFFF);
+
+    return CborNoError;
+}
+
+int suit_cbor_get_uint(const CborValue *it, unsigned *out)
+{
+#if 1
+    return suit_cbor_get_uint32(it, (uint32_t *)out);
+#else
+#error suit_cbor_get_uint only implemented for sizeof(unsigned) == 4
```

Comment:
Looks like leftover debugging code, so could be removed I guess

Reply:
yup